src/lavinmq/queue/stream_queue.cr

```diff
@@ -0,0 +1,228 @@
+        @size += 1
+        notify_empty(false) if was_empty
+        sp
+      end
+
+      def add_offset_header(msg, offset)
+        headers = msg.properties.headers || ::AMQP::Client::Arguments.new
+        headers["x-stream-offset"] = offset.as(AMQ::Protocol::Field)
+        msg.properties.headers = headers
+        msg
+      end
+
+      def offset_from_headers(headers)
+        if ht = headers.as?(AMQ::Protocol::Table)
+          ht["x-stream-offset"].as(Int64) # TODO: should not be i32, but cast fails
+        else
+          0_i64
+        end
+      end
+    end
+
+    # save message id / segment position
+    def publish(msg : Message) : Bool
+      return false if @state.closed?
+      reject_on_overflow(msg)
+      @msg_store_lock.synchronize do
+        @msg_store.push(msg)
+        @publish_count += 1
+      end
+      drop_overflow unless immediate_delivery?
+      true
+    rescue ex : MessageStore::Error
+      @log.error(exception: ex) { "Queue closed due to error" }
+      close
+      raise ex
+    end
+
+    def basic_get(no_ack, force = false, & : Envelope -> Nil) : Bool
+      return false if !@state.running? && (@state.paused? && !force)
+      @last_get_time = RoughTime.monotonic
+      @queue_expiration_ttl_change.try_send? nil
+      @get_count += 1
+      get(true) do |env|
+        yield env
+      end
+    end
+
+    # If nil is returned it means that the delivery limit is reached
+    def consume_get(consumer : Client::Channel::StreamConsumer, & : Envelope -> Nil) : Bool
+      get(consumer) do |env|
+        yield env
+        env.redelivered ? (@redeliver_count += 1) : (@deliver_count += 1)
+      end
+    end
+
+    # yield the next message in the ready queue
+    # returns true if a message was deliviered, false otherwise
+    # if we encouncer an unrecoverable ReadError, close queue
+    private def get(consumer : Client::Channel::StreamConsumer, & : Envelope -> Nil) : Bool
+      raise ClosedError.new if @closed
+      loop do # retry if msg expired or deliver limit hit
+        env = @msg_store_lock.synchronize { @msg_store.shift?(consumer) } || break
+        if has_expired?(env.message) # guarantee to not deliver expired messages
+          expire_msg(env, :expired)
+          next
+        end
+
+        sp = env.segment_position
+        begin
+          yield env # deliver the message
+        rescue ex   # requeue failed delivery
+          consumer.requeue(sp)
+          raise ex
+        end
+        return true
+      end
+      false
+    rescue ex : MessageStore::Error
+      @log.error(exception: ex) { "Queue closed due to error" }
+      close
+      raise ClosedError.new(cause: ex)
+    end
+
+    def ack(sp : SegmentPosition) : Nil
+      return if @deleted
+      true
+    end
+
+    # do nothing?
+    def reject(sp : SegmentPosition, requeue : Bool)
+      return if @deleted || @closed
+      true
+    end
+
+    def add_consumer(consumer : Client::Channel::StreamConsumer)
```

Comment:
how is this different from the base class implementation?

Reply:
Hmm, only removed some stuff that shouldn't be possible anyway (regarding exclusive consumer, priority consumers, etc). I think I can safely remove this method!